Fix SuperPMI implementation of `appendClassName`

BruceForstall opened the pull request:
Change #67135 added more and different calls to `appendClassName`, exposing
that it was not correctly implemented in SuperPMI and was implemented differently
in the EE and crossgen2. Fix this, especially, to handle the case of passing a zero length
to determine the correct size output buffer.

Also, add `eeTryGetClassSize` to wrap `getClassSize` and use it in two cases in the
importer that only use the output in JitDump, but was preventing a clean JitDump
replay of some collections.

Update and document the semantics of `appendClassName` API as follows.

Notable changes:
1. ppBuf argument is allowed to be nullptr, if `*pnBufLen` is zero.
2. If `*pnBufLen` is greater than zero, `*ppBuf` and `*pnBufLen` are
advanced by the number of characters actually written rather than the
number in the full class name string. This will differ in the case of
truncation due to a small output buffer. The advance does not include
the null terminator. Thus, `*ppBuf` will point at the null terminator
on return, and `*pnBufLen` will be at least `1`. This makes much more sense than incrementing
these based on the full name length, in case of truncation.
3. The EE will copy as much as possible (i.e., will truncate if necessary),
rather than crashing on a too-small output buffer.
4. Crossgen2 now advances `*ppBuf` and `*pnBufLen` as noted above.
5. SuperPMI implements these new semantics.

Change the JIT-EE GUID because the SuperPMI collection data structures were changed
and the API semantics changed.

BruceForstall committed:
Fix SuperPMI implementation of `appendClassName`

Change #67135 added more and different calls to `appendClassName`, exposing
that it was not correctly implemented in SuperPMI. Fix this, especially, to
handle the case of passing a zero length to determine the correct size output buffer.

Also, add `eeTryGetClassSize` to wrap `getClassSize` and use it in two cases in the
importer that only use the output in JitDump, but was preventing a clean JitDump
replay of some collections.

Change the JIT-EE GUID because the SuperPMI collection data structures were changed.

BruceForstall commented:
@kunalspathak @dotnet/jit-contrib PTAL

kunalspathak commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6450,58 +6450,111 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
+
+    if ((ppBuf != nullptr) && (*ppBuf != nullptr) && (*pnBufLen > 0) && (name != nullptr))
+    {
+        // Copy over as much as possible (note that the EE crashes is the buffer is too small).
```
```suggestion
        // Copy over as much as possible (note that the EE crashes if the buffer is too small).
```

kunalspathak commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6450,58 +6450,111 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
+        size_t sizeToCopy = min(*pnBufLen, nLen + 1 /* include null terminator */);
+        wcscpy_s((WCHAR*)*ppBuf, sizeToCopy, (const WCHAR*)name);
+        (*ppBuf)[sizeToCopy - 1] = 0;
+        (*ppBuf) += nLen;
```
Shouldn't this be `(*ppBuf) += sizeToCopy;` instead here and below?

kunalspathak replied:
Also, this method should return `sizeToCopy` instead of `nLen`? What is the contract here? A method doc will be helpful here.

BruceForstall replied:
Yeah, the function is terribly specified. I'll fix that (add documentation and make EE/crossgen2 match, and do something more reasonable when truncating)

kunalspathak reviewed (dismissed):
Few questions.

BruceForstall committed:
Update semantics of `appendClassName` API

Document the semantics more clearly.

Update EE and crossgen2 to conform to the same semantics.

Notable changes:
1. ppBuf argument is allowed to be nullptr, if `*pnBufLen` is zero.
2. If `*pnBufLen` is greater than zero, `*ppBuf` and `*pnBufLen` are
advanced by the number of characters actually written rather than the
number in the full class name string. This will differ in the case of
truncation due to a small output buffer. The advance does not include
the null terminator. Thus, `*ppBuf` will point at the null terminator
on return, and `*pnBufLen` will be at least `1`.
3. The EE will copy as much as possible (i.e., will truncate if necessary),
rather than crashing on a too-small output buffer.
4. Crossgen2 now advances `*ppBuf` and `*pnBufLen` as noted above.
5. SuperPMI implements these new semantics.

BruceForstall commented:
/azp run runtime-coreclr outerloop

BruceForstall commented:
@kunalspathak @MichalStrehovsky, all: I updated the documentation to be much more thorough, and updated the EE and crossgen2 (and SuperPMI) implementations to all be consistent with each other and the newly documented semantics.

MichalStrehovsky reviewed (approved):
The managed part looks good! Thanks for making the interface more sane!

kunalspathak commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -3369,14 +3369,24 @@ int CEEInfo::appendClassName(_Outptr_result_buffer_(*pnBufLen) char16_t** ppBuf,
+        // Copy as much as will fit.
+        WCHAR* pBuf = (WCHAR*)*ppBuf;
+        int nLenToCopy = min(*pnBufLen, nLen + /* null terminator */ 1);
+        for (int i = 0; i < nLenToCopy; i++)
```
Why not use `memcpy` instead?

kunalspathak replied:
Also shouldn't it be `i < (nLenToCopy - 1)`?

BruceForstall replied:
We could use memcpy; I wrote it this way so it looks more similar to the crossgen2 version, although they're not identical, so it's a minor point.

kunalspathak commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6450,58 +6450,116 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
+        // Copy as much as will fit.
+        char16_t* pBuf = *ppBuf;
+        int nLenToCopy = min(*pnBufLen, nLen + /* null terminator */ 1);
+        for (int i = 0; i < nLenToCopy; i++)
```
same here. We just need to do it until `i < nLenToCopy - 1`?

kunalspathak reviewed (commented):
Added few more comments. I might not be able to review further iterations, so go ahead and merge it if you decide to not fix any of my suggestions.

kunalspathak commented:
Also, looks like there are build failures that you might need to address.

BruceForstall committed:
Fix non-Windows build

Set `JIT_BUILD` when building SuperPMI. This disables contracts and some PAL debugging,
including the debugreturn.h debugging that was causing a build break.

Also, adjust a couple loops based on code review feedback.

BruceForstall commented:
/azp run runtime-coreclr outerloop

BruceForstall commented:
outerloop test failures are https://github.com/dotnet/runtime/issues/67355. runtime looks like https://github.com/dotnet/runtime/issues/60962
